File: rtipmd/rust/src/pes_exploration/md.rs
use std::fs::File;
use std::io::Write;
use crate::common::constants::*;
use crate::common::error::*;
use crate::pes_exploration::potential::*;
use crate::pes_exploration::rtip::*;
use crate::pes_exploration::traits::*;
use crate::pes_exploration::system::System;
use mpi::traits::*;
use ndarray::{Array2, Axis};
use ndarray_rand::RandomExt;
use ndarray_rand::rand_distr::Uniform;





impl<'a> RtipNVTMD for RepulsivePot<'a>
{
    fn rtip_nvt_md<C: Communicator, P: PES>(&self, comm: &C, real_pes: &P)
    {
        // Parallel processing
        let rank = comm.rank();
        let root_process = comm.process_at_rank(ROOT_RANK);

        // Output the local minimum structure to PDB output file, and output the header to RTIP output file
        if rank == ROOT_RANK
        {
            self.local_min.write_pdb(&self.str_output_file, true, 0);
            let mut rtip_output = File::create(&self.output_file).expect(&error_file("creating", &self.output_file));
            rtip_output.write_all(b"  step            time        rti_dist            temp             kin        pot_real        pot_rtip          f_real          f_rtip\n").expect(&error_file("writing", &self.output_file));
        }



        // Randomly obtain a nearby structure around self.local_min
        let mut s: System = self.local_min.clone();
        match &self.local_min.atom_add_pot
        {
            // If self.local_min.atom_add_pot has none value, disturb all the atoms
            None =>
            {
                let mut dcoord: Array2<f64> = Array2::random(s.coord.raw_dim(), Uniform::new(0.0, 1.0));
                let dcoord_p: &mut [f64] = dcoord.as_slice_mut().expect(&error_as_slice("dcoord"));
                root_process.broadcast_into(dcoord_p);
                dcoord = &dcoord - dcoord.mean_axis(Axis(0)).unwrap();
                let dcoord_norm: f64 = (&dcoord * &dcoord).sum().sqrt();
                s.coord += &(dcoord * (0.1/dcoord_norm));
            },

            // If self.local_min.atom_add_pot specifies the atoms to be added RTI potential, only disturb these atoms
            Some(atom_add_pot) =>
            {
                let mut dcoord: Array2<f64> = Array2::random((atom_add_pot.len(), 3), Uniform::new(0.0, 1.0));
                let dcoord_p: &mut [f64] = dcoord.as_slice_mut().expect(&error_as_slice("dcoord"));
                root_process.broadcast_into(dcoord_p);
                dcoord = &dcoord - dcoord.mean_axis(Axis(0)).unwrap();
                let dcoord_norm: f64 = (&dcoord * &dcoord).sum().sqrt();
                for i in 0..atom_add_pot.len()
                {
                    s.coord[[atom_add_pot[i], 0]] += dcoord[[i, 0]] * (0.1/dcoord_norm);
                    s.coord[[atom_add_pot[i], 1]] += dcoord[[i, 1]] * (0.1/dcoord_norm);
                    s.coord[[atom_add_pot[i], 2]] += dcoord[[i, 2]] * (0.1/dcoord_norm);
                }
            },
        }



        // Define the variables for NVT MD
        let mut pot_real: f64;
        let mut force_real: Array2<f64>;
        let mut f_real: f64;

        let mut local_min: System;                  // A fragment of the original local minimum containing the atoms to be added RTI potential
        let mut nearby_ts: Vec<System>;                 // A fragment of the original nearby TS containing the atoms to be added RTI potential
        let mut s_fragment: System;                 // A fragment of the current structure containing the atoms to be added RTI potential
        let mut rtip0_pes = match &self.local_min.atom_add_pot
        {
            // If self.local_min.atom_add_pot has none value, create the RTIP PES from the original local minimum and nearby TS
            None =>
            {
                // Initialize the fragments
                local_min = System
                {
                    natom: 0,
                    coord: Array2::zeros((0, 3)),
                    cell: None,
                    atom_type: None,
                    atom_add_pot: None,
                    mutable: None,
                    pot: 0.0,
                };
                nearby_ts = vec![local_min.clone(); self.nearby_ts.len()];
                s_fragment = local_min.clone();

                // Create the RTI potential from the original local minimum and nearby TS
                Rtip0PES
                {
                    local_min: &self.local_min,
                    nearby_ts: &self.nearby_ts,
                    a_min: 0.0,
                    a_ts: 0.0,
                    sigma_min: 0.0,
                    sigma_ts: vec![10.0; self.nearby_ts.len()],
                }
            },

            // If self.local_min.atom_add_pot specifies the atoms to be added RTI potential, create fragments for the local minimum and nearby TS
            Some(atom_add_pot) =>
            {
                // Initialize the fragments
                local_min = System
                {
                    natom: atom_add_pot.len(),
                    coord: Array2::zeros((atom_add_pot.len(), 3)),
                    cell: None,
                    atom_type: None,
                    atom_add_pot: None,
                    mutable: None,
                    pot: 0.0,
                };
                nearby_ts = vec![local_min.clone(); self.nearby_ts.len()];
                s_fragment = local_min.clone();

                // Copy the atomic coordinates from the original local minimum and nearby TS to the corresponding fragments
                for i in 0..atom_add_pot.len()
                {
                    local_min.coord[[i, 0]] = self.local_min.coord[[atom_add_pot[i], 0]];
                    local_min.coord[[i, 1]] = self.local_min.coord[[atom_add_pot[i], 1]];
                    local_min.coord[[i, 2]] = self.local_min.coord[[atom_add_pot[i], 2]];
                }
                for j in 0..self.nearby_ts.len()
                {
                    for i in 0..atom_add_pot.len()
                    {
                        nearby_ts[j].coord[[i, 0]] = self.nearby_ts[j].coord[[atom_add_pot[i], 0]];
                        nearby_ts[j].coord[[i, 1]] = self.nearby_ts[j].coord[[atom_add_pot[i], 1]];
                        nearby_ts[j].coord[[i, 2]] = self.nearby_ts[j].coord[[atom_add_pot[i], 2]];
                    }
                }

                // Create the RTIP PES from the fragments
                Rtip0PES
                {
                    local_min: &local_min,
                    nearby_ts: &nearby_ts,
                    a_min: 0.0,
                    a_ts: 0.0,
                    sigma_min: 0.0,
                    sigma_ts: vec![10.0; self.nearby_ts.len()],
                }
            },
        };

        let mut pot_rtip: f64;
        let mut force_rtip: Array2<f64> = Array2::zeros(s.coord.raw_dim());
        let mut force_rtip_fragment: Array2<f64>;
        let mut f_rtip: f64;

        let mut force_total: Array2<f64>;
        let dt: f64 = self.para.dt;
        let mut t: f64 = 0.0;
        let mut kin: f64;
        let mut temp: f64;                      // K
        let mut lambda: f64;
        let mut vel: Array2<f64> = Array2::zeros(s.coord.raw_dim());
        let mut acc: Array2<f64> = Array2::zeros(s.coord.raw_dim());
        let mut atom_mass: Vec<f64> = vec![0.0; s.natom];
        for i in 0..s.natom
        {
            atom_mass[i] = s.atom_type.as_ref().expect(&error_none_value("s.atom_type"))[i].get_mass();
        }



        // Perform the NVT MD iteractively
        for i in 1..(self.para.max_step+1)
        {
            // First step of leapfrog method
            t += dt;                    // fs
            vel += &(0.5 * dt * FEMTOSECOND_TO_AU * &acc);                  // A.U.
            s.coord += &(dt * FEMTOSECOND_TO_AU * &vel);                  // A.U.

            // Berendsen thermostat
            kin = 0.0;
            for j in 0..s.natom
            {
                kin += atom_mass[j] * ( vel[[j, 0]].powi(2) + vel[[j, 1]].powi(2) + vel[[j, 2]].powi(2) );          // Hartree
            }
            temp = kin * HARTREE_TO_JOULE / ( BOLTZMANN * 3.0 * (s.natom-1) as f64 );               // K
            if temp < 1.0
            {
                temp = 1.0;
            }
            lambda = ( 1.0 + (dt / self.para.tau) * (self.para.temp_bath / temp - 1.0) ).sqrt();

            // Calculate the real potential energy and atomic forces
            (pot_real, force_real) = real_pes.get_energy_force(&s);
            s.pot = pot_real;
            f_real = (&force_real * &force_real).sum().sqrt();

            // Calculate the RTIP potential energy and atomic forces
            match &self.local_min.atom_add_pot
            {
                // If self.local_min.atom_add_pot has none value, add the RTI potential on all the atoms
                None =>
                {
                    rtip0_pes.sigma_min = rti_dist(&self.local_min.coord, &s.coord);
                    match self.para.scale_ts_sigma
                    {
                        None =>
                        {
                            for j in 0..self.nearby_ts.len()
                            {
                                rtip0_pes.sigma_ts[j] = rti_dist(&self.nearby_ts[j].coord, &s.coord);
                            }
                        },
                        Some(scale_ts_sigma) =>
                        {
                            for j in 0..self.nearby_ts.len()
                            {
                                rtip0_pes.sigma_ts[j] = (0.5 * scale_ts_sigma) * rti_dist(&self.nearby_ts[j].coord, &self.local_min.coord);
                            }
                        },
                    }
                    rtip0_pes.a_min = self.para.a0 * (i as f64);
                    rtip0_pes.a_ts = rtip0_pes.a_min * self.para.scale_ts_a0;
                    (pot_rtip, force_rtip) = rtip0_pes.get_energy_force(&s);
                    f_rtip = (&force_rtip * &force_rtip).sum().sqrt();
                },

                // If self.local_min.atom_add_pot specifies the atoms to be add RTI potential, add the RTI potential on the fragments
                Some(atom_add_pot) =>
                {
                    for j in 0..atom_add_pot.len()
                    {
                        s_fragment.coord[[j, 0]] = s.coord[[atom_add_pot[j], 0]];
                        s_fragment.coord[[j, 1]] = s.coord[[atom_add_pot[j], 1]];
                        s_fragment.coord[[j, 2]] = s.coord[[atom_add_pot[j], 2]];
                    }
                    rtip0_pes.sigma_min = rti_dist(&local_min.coord, &s_fragment.coord);
                    match self.para.scale_ts_sigma
                    {
                        None =>
                        {
                            for j in 0..self.nearby_ts.len()
                            {
                                rtip0_pes.sigma_ts[j] = rti_dist(&nearby_ts[j].coord, &s_fragment.coord);
                            }
                        },
                        Some(scale_ts_sigma) =>
                        {
                            for j in 0..self.nearby_ts.len()
                            {
                                rtip0_pes.sigma_ts[j] = (0.5 * scale_ts_sigma) * rti_dist(&nearby_ts[j].coord, &local_min.coord);
                            }
                        },
                    }
                    rtip0_pes.a_min = self.para.a0 * (i as f64);
                    rtip0_pes.a_ts = rtip0_pes.a_min * self.para.scale_ts_a0;
                    (pot_rtip, force_rtip_fragment) = rtip0_pes.get_energy_force(&s_fragment);
                    f_rtip = (&force_rtip_fragment * &force_rtip_fragment).sum().sqrt();
                    for j in 0..atom_add_pot.len()
                    {
                        force_rtip[[atom_add_pot[j], 0]] = force_rtip_fragment[[j, 0]];
                        force_rtip[[atom_add_pot[j], 1]] = force_rtip_fragment[[j, 1]];
                        force_rtip[[atom_add_pot[j], 2]] = force_rtip_fragment[[j, 2]];
                    }
                },
            }

            // Calculate the atomic acceleration from the total force
            force_total = &force_real + &force_rtip;
            for j in 0..s.natom
            {
                acc[[j, 0]] = force_total[[j, 0]] / atom_mass[j];
                acc[[j, 1]] = force_total[[j, 1]] / atom_mass[j];
                acc[[j, 2]] = force_total[[j, 2]] / atom_mass[j];
            }

            // Second step of leapfrog method
            vel += &(0.5 * dt * FEMTOSECOND_TO_AU * &acc);              // A.U.
            vel *= lambda;

            // Calculate kinetic energy
            kin = 0.0;
            for j in 0..s.natom
            {
                kin += atom_mass[j] * ( vel[[j, 0]].powi(2) + vel[[j, 1]].powi(2) + vel[[j, 2]].powi(2) );              // Hartree
            }
            temp = kin * HARTREE_TO_JOULE / ( BOLTZMANN * 3.0 * (s.natom-1) as f64 );               // K
            kin *= 0.5;                 // Hartree

            // Output the information in this iterative step
            if rank == ROOT_RANK
            {
                if (i % self.para.print_step) == 0
                {
                    s.write_pdb(&self.str_output_file, false, i);
                }
                let mut rtip_output = File::options().append(true).open(&self.output_file).expect(&error_file("opening", &self.output_file));
                rtip_output.write_all(format!("{:6} {:15.8} {:15.8} {:15.8} {:15.8} {:15.8} {:15.8} {:15.8} {:15.8}\n", i, t, rtip0_pes.sigma_min, temp, kin, pot_real, pot_rtip, f_real, f_rtip).as_bytes()).expect(&error_file("writing", &self.output_file));
            }
        }
    }
}

File: rtipmd/jax/src/rtip_jax/workflows/md.py
"""RTIP-driven NVT MD, ported 1:1 from Rust `src/pes_exploration/md.rs`.

Two workflows are provided, matching the upstream implementations:

* ``repulsive_md``  – `RepulsivePot::rtip_nvt_md`; the RTIP amplitude grows
  linearly as ``a0 * step`` and pushes the system away from a local minimum
  (and optionally nearby transition states).
* ``evolution_md``  – `EvolutionPot::rtip_nvt_md`; an attractive RTIP pulls
  the molecules toward a virtual final state with coincident centroids.  The
  amplitude follows the three-state machine Increasing / Decreasing / Falling,
  driven by bond-variation detection and molecular-proximity checks.  This is
  the algorithm behind the formose-reaction RTIP-MD study.

The integrator is leapfrog with a Berendsen thermostat, exactly as in the Rust
implementation (units: fs, Bohr, Hartree).
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any, Sequence

import jax.numpy as jnp

from rtip_jax.config import Para
from rtip_jax.constants import (
    BOLTZMANN,
    FEMTOSECOND_TO_AU,
    HARTREE_TO_JOULE,
    Element,
    atomic_mass,
    atomic_radius,
    element_symbol,
)
from rtip_jax.core.rtip import Rtip0PES, rti_dist
from rtip_jax.io.pdb import write_pdb
from rtip_jax.pes.base import PES
from rtip_jax.pes.bias import EvolutionPot, RepulsivePot
from rtip_jax.system import (
    System,
    dist_mat_bohr,
    get_adj_mat,
    judge_adj_of_mol,
    judge_variation_of_bonding,
    split_into_mol,
)
from rtip_jax.workflows.pathway_sampling import force_norm, perturb_system


# ---------------------------------------------------------------------------
# Numeric helpers (Rust md.rs, exact unit conventions)
# ---------------------------------------------------------------------------


def atom_masses(system: System) -> Any:
    """Return atomic masses in atomic units for `system.atom_type`."""

    if system.atom_type is None:
        raise ValueError("system.atom_type is required for MD masses")
    return jnp.asarray([atomic_mass(element) for element in system.atom_type], dtype=jnp.float64)


def twice_kinetic_energy(velocity: Any, masses: Any) -> Any:
    """Return ``sum_i m_i |v_i|^2`` (Hartree), the value before ``*0.5``."""

    velocity = jnp.asarray(velocity, dtype=jnp.float64)
    masses = jnp.asarray(masses, dtype=jnp.float64)
    return jnp.sum(masses[:, jnp.newaxis] * velocity * velocity)


def kinetic_energy(velocity: Any, masses: Any) -> Any:
    """Return kinetic energy in Hartree."""

    return 0.5 * twice_kinetic_energy(velocity, masses)


def temperature(velocity: Any, masses: Any) -> Any:
    """Return instantaneous temperature in K (Rust formula)."""

    velocity = jnp.asarray(velocity, dtype=jnp.float64)
    if velocity.shape[0] <= 1:
        raise ValueError("temperature requires at least two atoms")
    return twice_kinetic_energy(velocity, masses) * HARTREE_TO_JOULE / (
        BOLTZMANN * 3.0 * float(velocity.shape[0] - 1)
    )


def berendsen_lambda(temp: Any, para: Para) -> Any:
    """Return the Berendsen thermostat scaling factor (temp clamped to >= 1 K)."""

    temp = jnp.maximum(jnp.asarray(temp, dtype=jnp.float64), 1.0)
    return jnp.sqrt(1.0 + (para.dt / para.tau) * (para.temp_bath / temp - 1.0))


def accelerations(force_total: Any, masses: Any) -> Any:
    """Return atomic accelerations from total force and masses."""

    force_total = jnp.asarray(force_total, dtype=jnp.float64)
    masses = jnp.asarray(masses, dtype=jnp.float64)
    return force_total / masses[:, jnp.newaxis]


def leapfrog_first(coord: Any, velocity: Any, acceleration: Any, para: Para) -> tuple[Any, Any]:
    """First Rust leapfrog half-step: update velocity and coordinates."""

    coord = jnp.asarray(coord, dtype=jnp.float64)
    velocity = jnp.asarray(velocity, dtype=jnp.float64)
    acceleration = jnp.asarray(acceleration, dtype=jnp.float64)
    velocity_half = velocity + 0.5 * para.dt * FEMTOSECOND_TO_AU * acceleration
    coord_new = coord + para.dt * FEMTOSECOND_TO_AU * velocity_half
    return coord_new, velocity_half


def leapfrog_second(velocity_half: Any, acceleration: Any, lambda_scale: Any, para: Para) -> Any:
    """Second Rust leapfrog half-step followed by thermostat scaling."""

    velocity_half = jnp.asarray(velocity_half, dtype=jnp.float64)
    acceleration = jnp.asarray(acceleration, dtype=jnp.float64)
    return (velocity_half + 0.5 * para.dt * FEMTOSECOND_TO_AU * acceleration) * lambda_scale


# ---------------------------------------------------------------------------
# Output records
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class MDStep:
    """One scalar-output row from an RTIP NVT MD workflow (Rust rtip.out)."""

    step: int
    time: float
    rti_dist: float
    temp: float
    kin: float
    pot_real: float
    pot_rtip: float
    f_real: float
    f_rtip: float
    rtip_status: str = ""


@dataclass(frozen=True)
class MDResult:
    """Final RTIP NVT MD state and scalar history."""

    system: System
    velocity: Any
    acceleration: Any
    history: tuple[MDStep, ...]


def _write_header(system: System, structure_file: str, output_file: str, write_outputs: bool) -> None:
    if not write_outputs:
        return
    Path(structure_file).parent.mkdir(parents=True, exist_ok=True)
    Path(output_file).parent.mkdir(parents=True, exist_ok=True)
    write_pdb(system, structure_file, create_new_file=True, step=0)
    Path(output_file).write_text(
        f"{'step':>6s} {'time_fs':>15s} {'rti_dist':>15s} {'temp_K':>15s} {'kin_Ha':>15s} "
        f"{'pot_real_Ha':>15s} {'pot_rtip_Ha':>15s} {'f_real':>15s} {'f_rtip':>15s} rtip_status\n"
    )


def _write_step(system: System, row: MDStep, structure_file: str, output_file: str, para: Para, write_outputs: bool) -> None:
    if not write_outputs:
        return
    if row.step % para.print_step == 0:
        write_pdb(system, structure_file, create_new_file=False, step=row.step)
    with Path(output_file).open("a") as handle:
        handle.write(
            f"{row.step:6d} {row.time:15.8f} {row.rti_dist:15.8f} {row.temp:15.8f} {row.kin:15.8f} "
            f"{row.pot_real:15.8f} {row.pot_rtip:15.8f} {row.f_real:15.8f} {row.f_rtip:15.8f} "
            f"{row.rtip_status}\n"
        )


def _validate_para(para: Para) -> None:
    if para.max_step < 0:
        raise ValueError("max_step must be non-negative")
    if para.print_step <= 0:
        raise ValueError("print_step must be positive")
    if para.dt <= 0.0:
        raise ValueError("dt must be positive")
    if para.tau <= 0.0:
        raise ValueError("tau must be positive")


def _atomic_radii(system: System) -> Any:
    """Return per-atom covalent radii in Bohr (Rust `get_atomic_radii`)."""

    import numpy as np

    if system.atom_type is None:
        raise ValueError("system.atom_type is required for bond detection")
    return np.asarray([atomic_radius(element) for element in system.atom_type], dtype=np.float64)


# ---------------------------------------------------------------------------
# RepulsivePot MD  (Rust `RepulsivePot::rtip_nvt_md`)
# ---------------------------------------------------------------------------


def repulsive_md(
    config: RepulsivePot,
    real_pes: PES,
    *,
    key: Any | None = None,
    perturb: bool = False,
    initial_velocity: Any | None = None,
    initial_acceleration: Any | None = None,
    write_outputs: bool = True,
) -> MDResult:
    """NVT MD driven by a repulsive RTIP with amplitude ``a0 * step``.

    The RTIP is built from the local minimum (and nearby TS, when provided)
    exactly as in Rust: ``a_min = a0 * i`` and ``a_ts = a_min * scale_ts_a0``.
    """

    _validate_para(config.para)
    para = config.para
    if perturb:
        if key is None:
            raise ValueError("key is required when perturb=True")
        s = perturb_system(config.local_min, key, atom_indices=config.local_min.atom_add_pot)
    else:
        s = config.local_min

    masses = atom_masses(s)
    velocity = jnp.zeros_like(s.coord) if initial_velocity is None else jnp.asarray(initial_velocity, dtype=jnp.float64)
    acceleration = (
        jnp.zeros_like(s.coord)
        if initial_acceleration is None
        else jnp.asarray(initial_acceleration, dtype=jnp.float64)
    )

    _write_header(config.local_min, config.str_output_file, config.output_file, write_outputs)
    history: list[MDStep] = []
    time = 0.0

    for step in range(1, para.max_step + 1):
        # First leapfrog half-step.
        time += para.dt
        coord_new, velocity_half = leapfrog_first(s.coord, velocity, acceleration, para)
        s = s.with_coord(coord_new)

        # Berendsen thermostat (half-step kinetic temperature).
        lambda_scale = berendsen_lambda(temperature(velocity_half, masses), para)

        # Real PES.
        pot_real, force_real = real_pes.get_energy_force(s)
        s = s.with_pot(float(pot_real))
        f_real = float(force_norm(force_real))

        # RTIP with linearly growing amplitude (a_min = a0 * step).
        indices = config.local_min.atom_add_pot
        sigma_min = float(rti_dist(_coords(config.local_min, indices), _coords(s, indices)))
        if para.scale_ts_sigma is None:
            sigma_ts = tuple(
                float(rti_dist(_coords(ts, indices), _coords(s, indices))) for ts in config.nearby_ts
            )
        else:
            sigma_ts = tuple(
                float(0.5 * para.scale_ts_sigma * rti_dist(_coords(ts, indices), _coords(config.local_min, indices)))
                for ts in config.nearby_ts
            )
        amplitude = para.a0 * float(step)
        bias_pes = Rtip0PES(
            local_min=config.local_min,
            nearby_ts=config.nearby_ts,
            a_min=amplitude,
            a_ts=amplitude * para.scale_ts_a0,
            sigma_min=sigma_min,
            sigma_ts=sigma_ts,
        )
        current = _with_atom_add_pot(s, indices)
        pot_bias, force_bias = bias_pes.get_energy_force(current)
        f_bias = float(force_norm(force_bias))

        # Forces -> acceleration; second leapfrog half-step + thermostat.
        force_total = jnp.asarray(force_real, dtype=jnp.float64) + jnp.asarray(force_bias, dtype=jnp.float64)
        acceleration = accelerations(force_total, masses)
        velocity = leapfrog_second(velocity_half, acceleration, lambda_scale, para)
        temp = float(temperature(velocity, masses))
        kin = float(kinetic_energy(velocity, masses))

        row = MDStep(
            step=step,
            time=time,
            rti_dist=sigma_min,
            temp=temp,
            kin=kin,
            pot_real=float(pot_real),
            pot_rtip=float(pot_bias),
            f_real=f_real,
            f_rtip=f_bias,
        )
        history.append(row)
        _write_step(s, row, config.str_output_file, config.output_file, para, write_outputs)

    return MDResult(system=s, velocity=velocity, acceleration=acceleration, history=tuple(history))


# ---------------------------------------------------------------------------
# EvolutionPot MD  (Rust `EvolutionPot::rtip_nvt_md`)
# ---------------------------------------------------------------------------


def _coords(system: System, indices: tuple[int, ...] | None) -> Any:
    if indices is None:
        return system.coord
    return system.coord[jnp.asarray(indices)]


def _with_atom_add_pot(system: System, indices: tuple[int, ...] | None) -> System:
    from dataclasses import replace

    return replace(system, atom_add_pot=indices)


def _final_state_coords(coord: Any, mol_index: tuple[tuple[int, ...], ...]) -> Any:
    """Virtual final state: move each molecule centroid toward the overall
    centroid by 0.01 of the current separation (Rust EvolutionPot)."""

    import numpy as np

    coord_np = np.asarray(coord, dtype=np.float64)
    o_all = coord_np.mean(axis=0)
    final_coord = coord_np.copy()
    for molecule in mol_index:
        if not molecule:
            continue
        o_i = coord_np[list(molecule)].mean(axis=0)
        final_coord[list(molecule)] += 0.01 * (o_all - o_i)
    return final_coord


def evolution_md(
    config: EvolutionPot,
    real_pes: PES,
    *,
    initial_velocity: Any | None = None,
    initial_acceleration: Any | None = None,
    write_outputs: bool = True,
) -> MDResult:
    """Formose-style evolution MD (Rust `EvolutionPot::rtip_nvt_md`).

    An attractive RTIP toward a coincident-centroid virtual final state drives
    the molecular assembly; bond-variation detection switches the RTIP from
    Increasing to Decreasing; the RTIP resets to Increasing once the depth
    returns to ``decreasing_bound`` of its peak.  Bond connectivity is
    re-derived at the beginning of every search round.
    """

    _validate_para(config.para)
    para = config.para
    s = config.initial_state
    indices = s.atom_add_pot

    if indices is not None:
        fragment = s.fragment(indices)
        system_for_bonds = fragment
        atomic_type: tuple[Element, ...] | None = fragment.atom_type
    else:
        system_for_bonds = s
        atomic_type = s.atom_type
    if atomic_type is None:
        raise ValueError("initial_state.atom_type is required for evolution MD")

    atomic_radii = _atomic_radii(system_for_bonds)
    dist_mat = dist_mat_bohr(system_for_bonds.coord)
    adj_mat = get_adj_mat(atomic_type, atomic_radii, dist_mat, 1.25, para.ignored_pair)
    mol_index = split_into_mol(atomic_radii, dist_mat, 1.25)

    masses = atom_masses(s)
    velocity = jnp.zeros_like(s.coord) if initial_velocity is None else jnp.asarray(initial_velocity, dtype=jnp.float64)
    acceleration = (
        jnp.zeros_like(s.coord)
        if initial_acceleration is None
        else jnp.asarray(initial_acceleration, dtype=jnp.float64)
    )

    _write_header(s, config.str_output_file, config.output_file, write_outputs)
    if write_outputs:
        Path(config.dec_output_file).parent.mkdir(parents=True, exist_ok=True)
        Path(config.dec_output_file).write_text("      begin_step        end_step\n")

    a_min = 0.0
    a_min_threshold = 0.0
    rtip_status = "Increasing"
    history: list[MDStep] = []
    time = 0.0

    for step in range(1, para.max_step + 1):
        # --- RTIP control (bond-variation detection) ---
        if rtip_status != "Decreasing" and judge_variation_of_bonding(
            atomic_radii, dist_mat, adj_mat, 1.0, 1.6
        ):
            a_min_threshold = a_min * para.decreasing_bound
            rtip_status = "Decreasing"
            if write_outputs:
                with Path(config.dec_output_file).open("a") as handle:
                    handle.write(f"{step:16d}")
        if rtip_status == "Decreasing" and a_min > a_min_threshold:
            adj_mat = get_adj_mat(atomic_type, atomic_radii, dist_mat, 1.25, para.ignored_pair)
            mol_index = split_into_mol(atomic_radii, dist_mat, 1.25)
            rtip_status = "Increasing"
            if write_outputs:
                with Path(config.dec_output_file).open("a") as handle:
                    handle.write(f"{step - 1:16d}\n")
        if para.split_step is not None and step % para.split_step == 0:
            mol_index = split_into_mol(atomic_radii, dist_mat, 1.25)

        # --- RTIP amplitude update ---
        if rtip_status == "Increasing":
            a_min -= para.a0
        elif rtip_status == "Decreasing":
            a_min += para.a0 * para.decreasing_multiple
        else:  # Falling
            a_min += para.a0

        # --- First leapfrog half-step ---
        time += para.dt
        coord_new, velocity_half = leapfrog_first(s.coord, velocity, acceleration, para)
        s = s.with_coord(coord_new)

        # --- Berendsen thermostat ---
        lambda_scale = berendsen_lambda(temperature(velocity_half, masses), para)

        # --- Real PES ---
        pot_real, force_real = real_pes.get_energy_force(s)
        s = s.with_pot(float(pot_real))
        f_real = float(force_norm(force_real))

        # --- Update distance matrix and the RTIP status ---
        current_bond_coord = _coords(s, indices)
        dist_mat = dist_mat_bohr(current_bond_coord)
        if rtip_status != "Decreasing":
            if judge_adj_of_mol(mol_index, atomic_radii, adj_mat, dist_mat, 1.2):
                rtip_status = "Falling"
            else:
                rtip_status = "Increasing"

        # --- Virtual final state (coincident molecular centroids) ---
        final_coord = _final_state_coords(current_bond_coord, mol_index)
        if indices is not None:
            subset_type = (
                tuple(s.atom_type[index] for index in indices) if s.atom_type is not None else None
            )
            final_state = System(
                coord=jnp.asarray(final_coord, dtype=jnp.float64),
                atom_type=subset_type,
            )
        else:
            final_state = System(coord=jnp.asarray(final_coord, dtype=jnp.float64))
        rtip_dist_value = float(rti_dist(final_state.coord, current_bond_coord))
        bias_pes = Rtip0PES(
            local_min=final_state,
            nearby_ts=(),
            a_min=a_min,
            a_ts=0.0,
            sigma_min=rtip_dist_value,
            sigma_ts=(),
        )
        pot_bias, force_bias = bias_pes.get_energy_force(_with_atom_add_pot(s, indices))
        f_bias = float(force_norm(force_bias))

        # --- Forces -> acceleration; second leapfrog half-step + thermostat ---
        force_total = jnp.asarray(force_real, dtype=jnp.float64) + jnp.asarray(force_bias, dtype=jnp.float64)
        acceleration = accelerations(force_total, masses)
        velocity = leapfrog_second(velocity_half, acceleration, lambda_scale, para)
        temp = float(temperature(velocity, masses))
        kin = float(kinetic_energy(velocity, masses))

        row = MDStep(
            step=step,
            time=time,
            rti_dist=rtip_dist_value,
            temp=temp,
            kin=kin,
            pot_real=float(pot_real),
            pot_rtip=float(pot_bias),
            f_real=f_real,
            f_rtip=f_bias,
            rtip_status=rtip_status,
        )
        history.append(row)
        _write_step(s, row, config.str_output_file, config.output_file, para, write_outputs)

    if write_outputs:
        with Path(config.dec_output_file).open("a") as handle:
            handle.write("\n")

    return MDResult(system=s, velocity=velocity, acceleration=acceleration, history=tuple(history))


__all__ = [
    "MDResult",
    "MDStep",
    "accelerations",
    "atom_masses",
    "berendsen_lambda",
    "evolution_md",
    "kinetic_energy",
    "leapfrog_first",
    "leapfrog_second",
    "repulsive_md",
    "temperature",
    "twice_kinetic_energy",
]
